src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -835,6 +855,43 @@ uct_ib_mlx5_devx_mem_reg_gva(uct_md_h uct_md, unsigned flags, uct_mem_h *memh_p)
     return status;
 }
 
+static ucs_status_t
+uct_ib_mlx5_devx_mem_window_reg(uct_md_h uct_md,
+                                const uct_md_mem_reg_params_t *params,
+                                uct_mem_h *memh_p)
+{
+    uct_ib_mlx5_md_t *md         = ucs_derived_of(uct_md, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_devx_mem_t *base =
+                    UCT_MD_MEM_REG_FIELD_VALUE(params, memh, FIELD_MEMH, NULL);
+    uct_ib_mlx5_devx_mem_t *memh;
+    ucs_status_t status;
+
+    if (ENABLE_PARAMS_CHECK && (base == NULL)) {
+        ucs_error("base memory handle is not provided for memory window");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    ucs_assertv(!(base->super.flags & UCT_IB_MEM_FLAG_MEM_WINDOW),
+                "memh=%p is already a memory window", base);
+
+    status = uct_ib_mlx5_devx_memh_clone(md, base, &memh);
+    if (status != UCS_OK) {
+        ucs_error("%s: failed to clone memory handle: %s",
+                  uct_ib_mlx5_dev_name(md), ucs_status_string(status));
+        return status;
+    }
+
+    /* Move indirect keys ownership from base to memory window */
+    memh->super.flags  |= UCT_IB_MEM_FLAG_MEM_WINDOW;
+    base->atomic_dvmr   = NULL;
+    base->atomic_rkey   = UCT_IB_INVALID_MKEY;
+    base->indirect_dvmr = NULL;
+    base->indirect_rkey = UCT_IB_INVALID_MKEY;
```

Comment:
why do we need to invalidate basic memh keys? I'd expect that this will break (at least performance) of some (at least atomics) operations on the base key

Reply:
I think it should not hinder performance, because base memh rkeys will be recreated on next mkey_pack, like is done in my test.
But you are right, strictly speaking it's NOT necessary to transfer rkeys ownership to the derived memh. It will work even without this transfer. I will remove this ownership transfer logic

The reason why I added this transfer is the following:
Currently it's UCT layer that decides whether to create rkeys (atomic + indirect), and conditions for them are different:
- **atomic** rkey is created if user requested it (with `UCT_MD_MEM_ACCESS_REMOTE_ATOMIC`), and hardware supports it. So it might be created no matter if invalidation is needed
- **indirect** rkey is created only when invalidation is needed
Now from UCP layer we assume that we will create a memory window when invalidation if needed from EP config. By this time **atomic** rkeys on base memh might be already created and I thought it's a good idea to move it to the memory window, to avoid re-creating this key per MW.

Reply:
I removed ownership transfer for now, let's consider this later